### PR TITLE
Add Array#each_with_index

### PR DESCRIPTION
**Pull request by giovannibonetti**
Add RBI definition for `Array#each_with_index` based on [the documentation on Apidock](https://apidock.com/ruby/Enumerator/each_with_index).

### Motivation
In my company we are very happy with our progress using Sorbet. We are gradually increasing type coverage and noticed Sorbet didn't know how to handle `Array#each_with_index` automatically.

### Test plan

Need help to guide me how to test this. I've tried:
- seeing how the tests for `each_index` were done, but [I found none](https://github.com/sorbet/sorbet/search?q=each_index&unscoped_q=each_index);
- seeing [other tests for Array methods](https://github.com/sorbet/sorbet/blob/master/test/testdata/rbi/array.rb), but none seemed to apply for this case, since we need to check the block arguments.

Is `T.assert_type!` useful in this case? 🤔

### CODE REVIEW

**Comment by elliottt** [comment]
Internal stripe build: http://go/builds/bui_HMCGWGc6435K5B

File: rbi/core/array.rbi [hunk 1]
```diff
@@ -1037,6 +1037,32 @@ class Array < Object
...
+  # value: b, index: 1
+  # value: c, index: 2
+  # ```
+  sig do
+    params(
+        blk: T.proc.params(arg0: Elem, arg1: Integer).returns(BasicObject),
+    )
+    .returns(T::Array[Elem])
+  end
+  sig {returns(T::Enumerator[Elem])}
```
Comment (elliottt) [inline rbi/core/array.rbi:1063]:
```suggestion
  sig {returns(T::Enumerator[[Elem,Integer]])}
```

I think that the return type on this overload is incorrect:
```
% irb
irb(main):001:0> [1, 2, 3].each_with_index.to_a
=> [[1, 0], [2, 1], [3, 2]]
```

Comment (giovannibonetti) [reply rbi/core/array.rbi]: Thank you for reviewing it! 👍

**Comment by elliottt** [comment]
Updated stripe build: http://go/builds/bui_HMCl5ElUzPxLXK

**Comment by elliottt** [comment]
Thanks!

**Pull request merged**